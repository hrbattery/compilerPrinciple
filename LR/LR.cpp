/*************************************************************************
	> File Name: LR.cpp
	  > Author: Netcan
	  > Blog: http://www.netcan666.com
	  > Mail: 1469709759@qq.com
	  > Created Time: 2016-10-17 一 21:03:17 CST

                                                    __----~~~~~~~~~~~------___
                                   .  .   ~~//====......          __--~ ~~
                   -.            \_|//     |||\\  ~~~~~~::::... /~
                ___-==_       _-~o~  \/    |||  \\            _/~~-
        __---~~~.==~||\=_    -_--~/_-~|-   |\\   \\        _/~
    _-~~     .=~    |  \\-_    '-~7  /-   /  ||    \      /
  .~       .~       |   \\ -_    /  /-   /   ||      \   /
 /  ____  /         |     \\ ~-_/  /|- _/   .||       \ /
 |~~    ~~|--~~~~--_ \     ~==-/   | \~--===~~        .\
          '         ~-|      /|    |-~\~~       __--~~
                      |-~~-_/ |    |   ~\_   _-~            /\
                           /  \     \__   \/~                \__
                       _--~ _/ | .-~~____--~-/                  ~~==.
                      ((->/~   '.|||' -_|    ~~-/ ,              . _||
                                 -_     ~\      ~~---l__i__i__i--~~_/
                                 _-~-__   ~)  \--______________--~~
                               //.-~~~-~_--~- |-------~~~~~~~~
                                      //.-~~~--\
						神兽保佑，代码无BUG!
 ************************************************************************/

#include "LR.h"
set<char>Item::Vn; // 全局静态变量
set<char>Item::Vt;
set<char>Item::Symbol;
const char*LR::actionStatStr[] = {
	"acc",
	"s",
	"r"
};

string Prod::displayStr() const{
	string p = string(1, noTerminal) + "->" + right.c_str();
	int i = 0;
	for(const auto& c:additionalVt)
		if(c != '#') p += string(1, i++==0?',':'|') + c; // #放到最后显示
	if(additionalVt.find('#') != additionalVt.end()) p += string(1, i++==0?',':'|') + "#";
	return p;
}

Prod::Prod(const string &in) {
	// printf("%s\n", in.c_str());
	noTerminal = in[0];
	right = cut(in, 3, in.length()); // A->X1X2X3X4
}

void Item::add(const string &prod) {
	if(prod.length() < 4) return;
	char noTerminal;
	if(Prod::cut(prod, 1, 3) == "->" && (isupper(prod[0]))) // A->...则noTerminal = A
		noTerminal = prod[0];
	else return;

	for(unsigned int i=0; i<prod.length(); ++i) { // 提取终结符、非终结符
		char c = prod[i];
		if(isupper(c)) {
			Vn.insert(c);
			Symbol.insert(c);
		}
		else if(c!='|' && !(c=='-' && prod[i+1] == '>' && ++i)) {
			Vt.insert(c);
			Symbol.insert(c);
		}
	}

	for(unsigned int i=3; i<prod.length(); ++i) { // 提取候选式
		unsigned int j;
		for(j=i+1; j<prod.length() && prod[j] != '|'; ++j);
		Prod p = Prod(string(1, noTerminal)+"->"+Prod::cut(prod, i, j));
		if(find(prods.begin(), prods.end(), p) == prods.end()) // 去重
			prods.push_back(p);
		i = j;
	}
}

void Item::display() const {
	for(const auto& prod: prods)
		cout << prod.displayStr() << endl;
}

void LR::add(const string &s) {
	G.add(s);
}

void LR::loadStr(const string &in) {
	inStr.push_back('#');
	status.push_back(0);
	for(int i = in.length() - 1; i>=0; --i)
		inStr.push_back(in[i]);
}

void LR::showStrStack() {
	for(vector<char>::reverse_iterator it = inStr.rbegin(); it != inStr.rend(); ++it)
		printf("%c", *it);
}


void LR::showStatusStack() {
	for(vector<int>::iterator it = status.begin(); it != status.end(); ++it) {
		if(*it < 10) printf(" %d ", *it);
		else printf(" <span class='underline'>%d</span> ", *it);
	}
}

void LR::showParseStack() {
	for(vector<char>::iterator it = parse.begin(); it != parse.end(); ++it)
		printf("%c", *it);
}

void LR::parser() {
	printf("\"parser\": [");

	bool success = false;
	int step = 0;
	while(! success) {
		// printf("%d\t", step);
		printf("%s{", step == 0?"\n":"\n, ");
		printf("\"statusStack\": \"");
		showStatusStack();
		printf("\", \"parseStack\": \"");
		showParseStack();
		printf("\", \"inStrStack\": \"");
		showStrStack();
		printf("\", \"action\": ");

		int sTop = status.size() - 1; // 栈顶
		int iTop = inStr.size() - 1;
		pair<int, char> p = make_pair(status[sTop], inStr[iTop]);
		if(ACTION.find(p) == ACTION.end())  // 出错！
			break;
	 	pair<actionStat, int> act = ACTION[p];
		if(act.first == SHIFT) { // 移进
			printf("\"SHIFT\"}");

			status.push_back(act.second);
			parse.push_back(inStr[iTop]);
			inStr.pop_back();
		} else if(act.first == REDUCE){
			Prod p = G.prods[act.second];
			printf("\"REDUCE %c->%s\"}", p.noTerminal, p.right.c_str());

			if(p.right != "@") // 空串，无需出栈，直接规约
				for(unsigned i=0; i<p.right.size(); ++i) {
					status.pop_back();
					parse.pop_back();
				}
			parse.push_back(p.noTerminal);
			status.push_back(GOTO[make_pair(status[status.size() - 1], p.noTerminal)]);
		} else if(act.first == ACCEPT) {
			success = true;
			printf("\"ACCEPT\"}");
		}
		++step;
	}
	if(! success)
		printf("\"ERROR\"}\n");

	printf("]\n");
}

Item LR::closure(Item I) {
	if(I.prods.size() == 0) return I;
	// size_t size = 0;
	// while(size != I.prods.size()) { // 当没有项目加入的时候
		// size = I.prods.size();
		// for(auto &prod: I.prods) { // bug
	for(size_t i = 0; i < I.prods.size(); ++i) { // 枚举I的产生式
		// puts("======================");
		// printf("i=%ld, size:%ld\n", i, I.prods.size());
		// I.display();
		Prod prod = I.prods[i];
		unsigned long pointLoc = 0;
		// printf("prod: %s\n", prod.displayStr().c_str());
		if((pointLoc = prod.right.find('.')) != string::npos && pointLoc != prod.right.length() - 1) { // 找到.，A->a.Bc,d
			char X = prod.right[pointLoc + 1];
			if(G.Vt.find(X) != G.Vt.end()) { // 终结符
				if(X == '@') // @特殊处理
					swap(I.prods[i].right[pointLoc], I.prods[i].right[pointLoc + 1]);
				continue;
			}

			string f = Prod::cut(prod.right, pointLoc+2, prod.right.length());
			// prod.display();
			// printf("f: %s\n", f.c_str());
			// printf("====================");
			set<char> ff = {};
			for(const auto& c: prod.additionalVt) {
				set<char> fs = first(f + c);
				ff.insert(fs.begin(), fs.end());
			}

			// if(ff == set<char>{'#'} && prod.noTerminal != EXTENSION_NOTERMINAL)  // 只含#，那么把Follow集加进来
			// ff.insert(FOLLOW[prod.noTerminal].begin(), FOLLOW[prod.noTerminal].end());


			for(vector<Prod>::iterator it = G.prods.begin(); it != G.prods.end(); ++it) {
				if(*it == X) { // 找到产生式
					Prod p = *it;
					if(p.right[0] == '@') { // 特殊处理.@ => @.
						p.right = '.' + p.right;
						swap(p.right[0], p.right[1]);
					} else
						p.right = '.' + p.right;

					vector<Prod>::iterator Iit = find(I.prods.begin(), I.prods.end(), p); // 找I中是否存在产生式
					if(Iit != I.prods.end())  // 找到
						Iit->additionalVt.insert(ff.begin(), ff.end());
					else {
						p.additionalVt.insert(ff.begin(), ff.end());
						I.prods.push_back(p);
					}
				}
			}
		}
	}
	// }
	return I;
}

Item LR::Goto(const Item& I, char X) {
	Item J;
	if(I.prods.size() == 0 || X == '@') return J; // 项目集为空或者@则返回空项目

	for(const auto& p: I.prods)  {// I中的每个项目
		string right = p.right;
		unsigned long pointLoc = right.find('.');
		if(right[pointLoc + 1] == X) {
			swap(right[pointLoc], right[pointLoc + 1]);
			J.prods.push_back(Prod(p.noTerminal, right, p.additionalVt));
		}
	}
	return closure(J);
}

void LR::items() {// 求项目集状态机DFA！!
	Item initial;
	initial.prods.push_back(Prod(EXTENSION_NOTERMINAL, '.' + string(1, G.prods[0].noTerminal), {'#'})); // 初值，^->.S,#
	C.push_back(closure(initial)); // 置C初值
	size_t size = 0;
	while(size != C.size()) { // 当没有项目集加入C中
		size = C.size();
		// for(auto &I: C) { // C的每个项目集，这样写有坑！！
		for(unsigned int i=0; i<C.size(); ++i) { // C的每个项目集
			Item I = C[i];
			for(const auto &X: G.Symbol) { // 每个文法Vt符号X
				Item next = Goto(I, X);
				if(next.prods.size() != 0) { // 不为空
					auto it = find(C.begin(), C.end(), next);
					if(it != C.end()) { // 找到
						GOTO[make_pair(i, X)] = it - C.begin();
					} else {
						C.push_back(next);
						GOTO[make_pair(i, X)] = C.size()-1;
					}
				}
			}
		}
	}
	// printf("size: %ld\n", C.size());
	// for(const auto& Item: C) {
		// printf("===%ld===\n", &Item - &C[0]);
		// Item.display();
	// }
}

void LR::build() { // 构造Action、GOTO表
	// follow();
	items();
	for(unsigned int i=0; i<C.size(); ++i) { // 逐个项目集
		const Item & item = C[i];
		for(const auto& prod: item.prods) { // 逐个项目
			unsigned long pointLoc = prod.right.find('.');
			if(pointLoc < prod.right.length() - 1) { // 不是最后一个
				char X = prod.right[pointLoc + 1];
				if(G.Vt.find(X) != G.Vt.end() && GOTO.find(make_pair(i, X)) != GOTO.end()) { // 终结符
					int j = GOTO[make_pair(i, X)];
					ACTION[make_pair(i, X)] = make_pair(SHIFT, j);
				}
			} else {
				if(prod == Prod(EXTENSION_NOTERMINAL, string(1, G.prods[0].noTerminal)+'.', {}) && prod.additionalVt == set<char>({'#'}))  // S'->S.,# acction[i, #] = acc
					ACTION[make_pair(i, '#')] = make_pair(ACCEPT, 0);
				else if(prod.noTerminal != EXTENSION_NOTERMINAL) {
					string right = prod.right;
					right.erase(pointLoc, 1); // 删除.
					for(const auto& X: prod.additionalVt) { // A->a.,b，枚举b
						vector<Prod>::iterator it = find(G.prods.begin(), G.prods.end(), Prod(prod.noTerminal, right, set<char>{}));
						if(it != G.prods.end())  // 找到了
							ACTION[make_pair(i, X)] = make_pair(REDUCE, it - G.prods.begin());
					}
				}
			}
		}
	}
	if(G.Vt.find('@') != G.Vt.end()) { // 删除@，移进#
		G.Vt.erase(G.Vt.find('@'));
		G.Symbol.erase(G.Symbol.find('@'));
	}
	G.Vt.insert('#');
	G.Symbol.insert('#');
}

void LR::showTable() {
	printf("\"parseTable\": {\n");
	printf("\"Vt\": [");

	for(const auto & X: G.Vt) {
		if(X != '#')
			printf("\"%c\", ", X);
	}
	printf("\"%c\"", '#'); // #放到最后一列显示，美观
	printf("],\n\"Vn\": [");

	int firstComma = 0; // 处理第一个逗号
	for(const auto & X: G.Vn)
		printf("%s\"%c\"", firstComma++ == 0?" ":", ",X);
	printf("],\n\"Body\": [\n");

	for(unsigned int i=0; i<C.size(); ++i) {
		printf("%s[", i == 0?" ":", ");
		int firstComma = 0;
		for(const auto & X: G.Vt) {
			if(X != '#') {
				pair<int, char> p = make_pair(i, X);
				if(ACTION.find(p) != ACTION.end()) {
					pair<actionStat, int> res = ACTION[p];
					printf("%s\"%s%d\"", firstComma++ == 0?" ":", ", actionStatStr[res.first], res.second);
				}
				else printf("%s\"\"", firstComma++ == 0?" ":", ");
			}
		}

		pair<int, char> p = make_pair(i, '#');
		if(ACTION.find(p) != ACTION.end()) {
			pair<actionStat, int> res = ACTION[p];
			printf("%s\"%s", firstComma++ == 0?" ":", ", actionStatStr[res.first]);
			if(res.first != ACCEPT)
				printf("%d\"", res.second);
			else printf("\"");
		}
		else printf("%s\"\"", firstComma++ == 0?" ":", ");

		for(const auto & X: G.Vn) {
			pair<int, char> p = make_pair(i, X);
			if(GOTO.find(p) != GOTO.end())
				printf("%s%d", firstComma++ == 0?" ":", ", GOTO[make_pair(i, X)]);
			else printf("%s\"\"", firstComma++ == 0?" ":", ");
		}
		puts("]");
	}

	printf("]}\n");
}

set<char> LR::first(const string &s) { // s不为产生式！
	if(s.length() == 0)
		return set<char>({'@'});
	else if(s.length() == 1) {
		if(G.Vt.find(s[0]) != G.Vt.end() || s[0] == '#') // 终结符
			return set<char>({s[0]});
		else
			if(FIRST[s[0]].size() != 0) return FIRST[s[0]];
			else {
				for(vector<Prod>::iterator it = G.prods.begin(); it != G.prods.end(); ++it)
					if(*it == s[0]) {
						// 防止直接左递归
						size_t xPos = it->right.find(it->noTerminal);
						// printf("prod: %s right: %s\n", it->displayStr().c_str(), it->right.c_str());
						if(xPos != string::npos) { // 找到X->aXb
							if(xPos == 0) continue; // X->Xb
							else { // X->aXb
								string a = Prod::cut(it->right, 0, xPos);
								if(first(a) == set<char>{'@'}) continue;
							}
						}
						set<char> f = first(it->right);
						FIRST[s[0]].insert(f.begin(), f.end());
					}
				// printf("first(%s) = ", s.c_str());
				// for(auto c: FIRST[s[0]])
					// printf("%c, ", c);
				// puts("");

				return FIRST[s[0]];
			}
	} else { // first(X1X2X3X4)...
		set<char> ret;
		for(unsigned int i=0; i<s.length(); ++i) {
			set<char> f = first(string(1, s[i])); // 逐个符号求first(Xi)集
			if(f.find('@') != f.end() && s.length() - 1 != i) { // 发现@
				f.erase(f.find('@')); // 减去@
				ret.insert(f.begin(), f.end()); // 放入first集合
			} else { // 无@或者最后一个Xi，则不需要求下去了
				ret.insert(f.begin(), f.end());
				break;
			}
		}
		return ret;
	}
}


void LR::follow() {
	FOLLOW[G.prods[0].noTerminal].insert('#'); // 开始符号放'#'
	for(auto pp: G.prods) { // 直到follow(X)不在增大
		unsigned int size = 0;
		while(size != FOLLOW[pp.noTerminal].size()) {
			size = FOLLOW[pp.noTerminal].size();
			for(auto prod: G.prods) { // 求出所有非终结符的follow集合
				char X = prod.noTerminal;
				for(auto p: G.prods) {// 求出X的follow集合
					string s = p.right;
					unsigned long loc = 0;
					if((loc = s.find(X)) != string::npos) { // 找到非终结符X
						set<char> f = first(string(s.begin() + loc + 1, s.end())); // 求first(b)
						FOLLOW[X].insert(f.begin(), f.end()); // 加入到follow(X)中
						if(f.find('@') != f.end()) {// 找到@
							FOLLOW[X].erase(FOLLOW[X].find('@')); // 删除@
							set<char> fw = FOLLOW[p.noTerminal]; // 把follow(A)放入follow(X)
							FOLLOW[X].insert(fw.begin(), fw.end());
						}
					}
				}
			}
		}
	}
}

void LR::debug() {
	string in;
	while(cin >> in && in != "#")
		add(in);

	puts("=====Proj:======");
	for(auto pro: G.prods)
		printf("%s\n", (string(1, pro.noTerminal) + "->" + pro.right).c_str());
	puts("=====Vt:======");
	for(auto vt: G.Vt)
		printf("%c, ", vt);
	puts("\n=====Vn:======");
	for(auto vn: G.Vn)
		printf("%c, ", vn);
	// puts("\n=====FIRST:======");
	// set<char> f = first("S");
	// for(auto c: f)
		// printf("%c, ", c);
	puts("");
	// puts("\n=====Item:======");
	// Item I;
	// I.prods.push_back(Prod('B', "b.B", {'b', 'a'}));
	// I = closure(I);
	// for(auto prod: I.prods)
		// prod.display();
	build();
	showGrammar();

	in = "";
	cin >> in;
	if(in == "#" || in.size() == 0) return;
	loadStr(in);
	parser();
	puts("");
}

void LR::showGrammar() {
	printf("\"Grammar\": [\n");
	for(const auto & p: G.prods)
		printf("%s\"%c->%s\"", (&p - &G.prods[0]) == 0?" ":", ", p.noTerminal, p.right.c_str());
	printf("\n]\n");
}

string Prod::replaceAll(const string &in, const string from, const string to) {
	size_t replacePos = in.find(from);
	string res = in;
	if(replacePos != string::npos)
		res.replace(replacePos, from.length(), to);
	return res;
}

void LR::drawGraph() {
	printf("\"Graph\": {");
	// 图的全部信息
	printf("\"All\": \"");
	printf("digraph all{"
			"node [shape=box style=filled];"
		  );
	// 文法信息
	printf("Grammar[style=rounded label=\\\"Grammar\\n");
	for(const auto &prod: G.prods)
		printf("%s\\n", prod.displayStr().c_str());
	printf("\\\" ];\\n");

	// 画节点
	for(const auto &I: C) { // 遍历项目集
		int i = &I - &C[0];
		printf("I%d[label=\\\"I%d\\n", i, i);
		for(const auto &p: I.prods) { // 列出项目
			string res = p.displayStr();
			if(res.find('^') != string::npos)
				res = Prod::replaceAll(res, "^", string(1, G.prods[0].noTerminal)+"'");
			printf("%s\\n", res.c_str());
		}
		printf("\\\" ];\\n");
	}

	// 画边
	for(const auto &link: GOTO) {
		int i = link.first.first;
		string X = string(1, link.first.second);
		int j = link.second;
		printf("I%d -> I%d[label=\\\"%s\\\"];", i, j, X.c_str());
	}
	printf("Grammar -> I0[style=invis];");
	printf("}\"\n, ");

	// 图的简要信息
	printf("\"Simple\": \"");
	printf("digraph simple {"
			"node [shape = circle style=filled];"
		  );
	// 文法信息
	printf("Grammar[shape=box style=rounded label=\\\"Grammar\\n");
	for(const auto &prod: G.prods)
		printf("%s\\n", prod.displayStr().c_str());
	printf("\\\" ];\\n");

	// 画节点
	for(const auto &I: C) { // 遍历项目集
		int i = &I - &C[0];
		printf("I%d[tooltip=\\\"I%d\\n", i, i);
		for(const auto &p: I.prods) { // 列出项目
			string res = p.displayStr();
			if(res.find('^') != string::npos)
				res = Prod::replaceAll(res, "^", string(1, G.prods[0].noTerminal)+"'");
			printf("%s\\n", res.c_str());
		}
		printf("\\\" ];\\n");
	}
	// 画边
	for(const auto &link: GOTO) {
		int i = link.first.first;
		string X = string(1, link.first.second);
		int j = link.second;
		printf("I%d -> I%d[label=\\\"%s\\\"];", i, j, X.c_str());
	}
	printf("Grammar -> I0[style=invis];");

	printf("}\"");
	printf("}");

}

void LR::generateDot() {
	printf("digraph all{\n"
			"node [shape=box style=filled];\n"
		  );
	// 文法信息
	printf("Grammar[style=rounded label=\"Grammar\n");
	for(const auto &prod: G.prods)
		printf("%s\n", prod.displayStr().c_str());
	printf("\" ];\n");

	// 画节点
	for(const auto &I: C) { // 遍历项目集
		int i = &I - &C[0];
		printf("I%d[label=\"I%d\n", i, i);
		for(const auto &p: I.prods) { // 列出项目
			string res = p.displayStr();
			if(res.find('^') != string::npos)
				res = Prod::replaceAll(res, "^", string(1, G.prods[0].noTerminal)+"'");
			printf("%s\n", res.c_str());
		}
		printf("\" ];\n");
	}

	// 画边
	for(const auto &link: GOTO) {
		int i = link.first.first;
		string X = string(1, link.first.second);
		int j = link.second;
		printf("I%d -> I%d[label=\"%s\"];\n", i, j, X.c_str());
	}
	printf("Grammar -> I0[style=invis];");
	printf("\n}\n");
}

void LR::run() {
	string in;
	while(cin >> in && in != "#")
		add(in);
	in = "";

	printf("{");
	showGrammar();
	printf(",");

	build();
	showTable();
	printf(",");

	drawGraph();
	printf(",\n");

	cin >> in;
	if(in == "#" || in.size() == 0) return;
	loadStr(in);
	parser();
	printf("}\n");
}

#ifndef _GENERATE_DOT_
int main() {
	LR lr;
	lr.run();

	return 0;
}
#endif


