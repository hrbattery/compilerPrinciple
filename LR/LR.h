/*************************************************************************
	> File Name: LR.h
	  > Author: Netcan
	  > Blog: http://www.netcan666.com
	  > Mail: 1469709759@qq.com
	  > Created Time: 2016-10-19 三 08:36:05 CST

                                                    __----~~~~~~~~~~~------___
                                   .  .   ~~//====......          __--~ ~~
                   -.            \_|//     |||\\  ~~~~~~::::... /~
                ___-==_       _-~o~  \/    |||  \\            _/~~-
        __---~~~.==~||\=_    -_--~/_-~|-   |\\   \\        _/~
    _-~~     .=~    |  \\-_    '-~7  /-   /  ||    \      /
  .~       .~       |   \\ -_    /  /-   /   ||      \   /
 /  ____  /         |     \\ ~-_/  /|- _/   .||       \ /
 |~~    ~~|--~~~~--_ \     ~==-/   | \~--===~~        .\
          '         ~-|      /|    |-~\~~       __--~~
                      |-~~-_/ |    |   ~\_   _-~            /\
                           /  \     \__   \/~                \__
                       _--~ _/ | .-~~____--~-/                  ~~==.
                      ((->/~   '.|||' -_|    ~~-/ ,              . _||
                                 -_     ~\      ~~---l__i__i__i--~~_/
                                 _-~-__   ~)  \--______________--~~
                               //.-~~~-~_--~- |-------~~~~~~~~
                                      //.-~~~--\
						神兽保佑，代码无BUG!
 ************************************************************************/

#ifndef _LR_H
#define _LR_H

#include <iostream>
#include <ctype.h>
#include <vector>
#include <string>
#include <algorithm>
#include <map>
#include <set>
using namespace std;
#define EXTENSION_NOTERMINAL '^'

class Prod { // 这里是存放形如X->abc的形式，不存在多个候选式
	friend class Item;
	friend class LR;
	private:
		char noTerminal; // 产生式左部非终结符名字
		string right; // 产生式右部
		set<char> additionalVt; // 附加终结符
		friend bool operator == (const Prod &a, const Prod &b) {
			return a.noTerminal == b.noTerminal && a.right == b.right;
		}
		friend bool operator == (const Prod &a, char c) {
			return a.noTerminal == c;
		}

	public:
		static string cut(const string &in, int i, int j) {
			return string(in.begin() + i, in.begin() + j);
		}
		static string replaceAll(const string &in, const string from, const string to);
		string displayStr() const;
		Prod(const string &in);
		Prod(const char &noTerminal, const string& right, const set<char>& additionalVt):
			noTerminal(noTerminal), right(right), additionalVt(additionalVt) {}
};

class Item { // 项目集
	friend class LR;
	private:
		vector<Prod> prods; // 项目集
		static set<char> Vn; // 非终结符
		static set<char> Vt; // 终结符
		static set<char> Symbol; // 所有符号
		friend bool operator == (const Item &a, const Item &b) {
			if(a.prods.size() != b.prods.size()) return false;
			else {
				for(const auto& p: a.prods) {// 选择a的每个产生式
					auto it = find(b.prods.begin(), b.prods.end(), p);
					if(it == b.prods.end())  // 找不到
						return false;
					else  {// 找到的话判断附加终结符是否都相等
						if(p.additionalVt != it->additionalVt)
							return false;
					}
				}
				return true;
			}
		}
	public:
		void add(const string &prod);
		void display() const;
};

class LR {
	private:
		Item G; // 文法G
		enum actionStat{
			ACCEPT=0,
			SHIFT,
			REDUCE,
		};
		static const char *actionStatStr[];

		vector<Item> C; // 项目集规范族
		map<pair<int, char>, int> GOTO; // goto数组，项目集<int, int>=char
		map<pair<int, char>, pair<actionStat, int> > ACTION; // Action数组，Action[(i, a)]=(s|r)j
		map<char, set<char> > FIRST; // first集
		map<char, set<char> > FOLLOW; // follow集
		set<char> first(const string &s); // 求first集
		void follow();
		vector<char> inStr; // 输入串/栈
		vector<int> status; // 状态栈
		vector<char> parse; // 分析栈
		Item closure(Item I); // 求该项目的闭包
		Item Goto(const Item& I, char X); // 求I经过X到达的项目集
		void items(); // 求项目集状态机DFA！!
		void showStrStack(); // 显示输入栈
		void showStatusStack();
		void showParseStack();
	public:
		void add(const string &s); // 添加产生式
		void build(); // 构造Action、GOTO表
		void showTable(); // 打印LR分析表！
		void debug();
		void loadStr(const string &in); // 读取输入串
		void parser(); // LR(1)分析
		void showGrammar(); // 显示输入的文法产生式
		void drawGraph(); // 画出DFA！
		void generateDot();
		void run();
};

#endif
