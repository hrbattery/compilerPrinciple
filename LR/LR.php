<?php
header('Content-Type: application/json');

$data = $_POST['data'] . "\n#\n#";
$descriptorspec = array(
   0 => array("pipe", "r"),  // stdin is a pipe that the child will read from
   1 => array("pipe", "w"),  // stdout is a pipe that the child will write to
);

if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN')
	$process = proc_open('LR', $descriptorspec, $pipes);
else
	$process = proc_open('./LR', $descriptorspec, $pipes);

if(is_resource($process)) {

	fwrite($pipes[0], $data);
	fclose($pipes[0]);

	echo stream_get_contents($pipes[1]);
	fclose($pipes[1]);
	proc_close($process);
}


