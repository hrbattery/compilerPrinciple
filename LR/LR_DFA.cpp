/*************************************************************************
	> File Name: LR_DFA.cpp
	  > Author: Netcan
	  > Blog: http://www.netcan666.com
	  > Mail: 1469709759@qq.com
	  > Created Time: 2016-10-21 五 23:08:06 CST

                                                    __----~~~~~~~~~~~------___
                                   .  .   ~~//====......          __--~ ~~
                   -.            \_|//     |||\\  ~~~~~~::::... /~
                ___-==_       _-~o~  \/    |||  \\            _/~~-
        __---~~~.==~||\=_    -_--~/_-~|-   |\\   \\        _/~
    _-~~     .=~    |  \\-_    '-~7  /-   /  ||    \      /
  .~       .~       |   \\ -_    /  /-   /   ||      \   /
 /  ____  /         |     \\ ~-_/  /|- _/   .||       \ /
 |~~    ~~|--~~~~--_ \     ~==-/   | \~--===~~        .\
          '         ~-|      /|    |-~\~~       __--~~
                      |-~~-_/ |    |   ~\_   _-~            /\
                           /  \     \__   \/~                \__
                       _--~ _/ | .-~~____--~-/                  ~~==.
                      ((->/~   '.|||' -_|    ~~-/ ,              . _||
                                 -_     ~\      ~~---l__i__i__i--~~_/
                                 _-~-__   ~)  \--______________--~~
                               //.-~~~-~_--~- |-------~~~~~~~~
                                      //.-~~~--\
						神兽保佑，代码无BUG!
 ************************************************************************/

#include "LR.h"

int main() {
	LR lr;
	string in;
	while(cin >> in && in != "#")
		lr.add(in);
	in = "";
	cin >> in;
	if(in == "#" || in.size() == 0) return -1;
	lr.loadStr(in);
	lr.build();
	lr.generateDot();
}
