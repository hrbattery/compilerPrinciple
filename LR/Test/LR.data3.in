S->BB
B->bB|a
#
baba
