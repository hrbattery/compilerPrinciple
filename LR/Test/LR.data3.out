{"Grammar": [
 "S->BB", "B->bB", "B->a"
]
,"parseTable": {
"Vt": ["a", "b", "#"],
"Vn": [ "B", "S"],
"Body": [
 [ "s3", "s4", "", 1, 2]
, [ "s6", "s7", "", 5, ""]
, [ "", "", "acc", "", ""]
, [ "r2", "r2", "", "", ""]
, [ "s3", "s4", "", 8, ""]
, [ "", "", "r0", "", ""]
, [ "", "", "r2", "", ""]
, [ "s6", "s7", "", 9, ""]
, [ "r1", "r1", "", "", ""]
, [ "", "", "r1", "", ""]
]}
,"Graph": {"All": "digraph all{node [shape=box style=filled];Grammar[style=rounded label=\"Grammar\nS->BB\nB->bB\nB->a\n\" ];\nI0[label=\"I0\nS'->.S,#\nS->.BB,#\nB->.bB,a|b\nB->.a,a|b\n\" ];\nI1[label=\"I1\nS->B.B,#\nB->.bB,#\nB->.a,#\n\" ];\nI2[label=\"I2\nS'->S.,#\n\" ];\nI3[label=\"I3\nB->a.,a|b\n\" ];\nI4[label=\"I4\nB->b.B,a|b\nB->.bB,a|b\nB->.a,a|b\n\" ];\nI5[label=\"I5\nS->BB.,#\n\" ];\nI6[label=\"I6\nB->a.,#\n\" ];\nI7[label=\"I7\nB->b.B,#\nB->.bB,#\nB->.a,#\n\" ];\nI8[label=\"I8\nB->bB.,a|b\n\" ];\nI9[label=\"I9\nB->bB.,#\n\" ];\nI0 -> I1[label=\"B\"];I0 -> I2[label=\"S\"];I0 -> I3[label=\"a\"];I0 -> I4[label=\"b\"];I1 -> I5[label=\"B\"];I1 -> I6[label=\"a\"];I1 -> I7[label=\"b\"];I4 -> I8[label=\"B\"];I4 -> I3[label=\"a\"];I4 -> I4[label=\"b\"];I7 -> I9[label=\"B\"];I7 -> I6[label=\"a\"];I7 -> I7[label=\"b\"];Grammar -> I0[style=invis];}"
, "Simple": "digraph simple {node [shape = circle style=filled];Grammar[shape=box style=rounded label=\"Grammar\nS->BB\nB->bB\nB->a\n\" ];\nI0[tooltip=\"I0\nS'->.S,#\nS->.BB,#\nB->.bB,a|b\nB->.a,a|b\n\" ];\nI1[tooltip=\"I1\nS->B.B,#\nB->.bB,#\nB->.a,#\n\" ];\nI2[tooltip=\"I2\nS'->S.,#\n\" ];\nI3[tooltip=\"I3\nB->a.,a|b\n\" ];\nI4[tooltip=\"I4\nB->b.B,a|b\nB->.bB,a|b\nB->.a,a|b\n\" ];\nI5[tooltip=\"I5\nS->BB.,#\n\" ];\nI6[tooltip=\"I6\nB->a.,#\n\" ];\nI7[tooltip=\"I7\nB->b.B,#\nB->.bB,#\nB->.a,#\n\" ];\nI8[tooltip=\"I8\nB->bB.,a|b\n\" ];\nI9[tooltip=\"I9\nB->bB.,#\n\" ];\nI0 -> I1[label=\"B\"];I0 -> I2[label=\"S\"];I0 -> I3[label=\"a\"];I0 -> I4[label=\"b\"];I1 -> I5[label=\"B\"];I1 -> I6[label=\"a\"];I1 -> I7[label=\"b\"];I4 -> I8[label=\"B\"];I4 -> I3[label=\"a\"];I4 -> I4[label=\"b\"];I7 -> I9[label=\"B\"];I7 -> I6[label=\"a\"];I7 -> I7[label=\"b\"];Grammar -> I0[style=invis];}"},
"parser": [
{"statusStack": " 0 ", "parseStack": "", "inStrStack": "baba#", "action": "SHIFT"}
, {"statusStack": " 0  4 ", "parseStack": "b", "inStrStack": "aba#", "action": "SHIFT"}
, {"statusStack": " 0  4  3 ", "parseStack": "ba", "inStrStack": "ba#", "action": "REDUCE B->a"}
, {"statusStack": " 0  4  8 ", "parseStack": "bB", "inStrStack": "ba#", "action": "REDUCE B->bB"}
, {"statusStack": " 0  1 ", "parseStack": "B", "inStrStack": "ba#", "action": "SHIFT"}
, {"statusStack": " 0  1  7 ", "parseStack": "Bb", "inStrStack": "a#", "action": "SHIFT"}
, {"statusStack": " 0  1  7  6 ", "parseStack": "Bba", "inStrStack": "#", "action": "REDUCE B->a"}
, {"statusStack": " 0  1  7  9 ", "parseStack": "BbB", "inStrStack": "#", "action": "REDUCE B->bB"}
, {"statusStack": " 0  1  5 ", "parseStack": "BB", "inStrStack": "#", "action": "REDUCE S->BB"}
, {"statusStack": " 0  2 ", "parseStack": "S", "inStrStack": "#", "action": "ACCEPT"}]
}
